rsc/src/test/java/com/cloudera/livy/rsc/TestSparkClient.java

```diff
@@ -346,6 +338,65 @@ void config(Properties conf) {
   }
 
   @Test
+  public void testKillServerWhileSparkSubmitIsRunning() throws Exception {
+    Properties conf = createConf(true);
+    LivyClient client = null;
+    PipedInputStream stubStream = new PipedInputStream(new PipedOutputStream());
+    try {
+      Process mockSparkSubmit = mock(Process.class);
+      when(mockSparkSubmit.getInputStream()).thenReturn(stubStream);
+      when(mockSparkSubmit.getErrorStream()).thenReturn(stubStream);
+
+      // Block waitFor until process.destroy is called.
+      final CountDownLatch waitForCalled = new CountDownLatch(1);
+      when(mockSparkSubmit.waitFor()).thenAnswer(new Answer<Void>() {
+        @Override
+        public Void answer(InvocationOnMock invocation) throws Throwable {
+          waitForCalled.await();
+          return null;
+        }
+      });
+
+      ContextLauncher.mockSparkSubmit = mockSparkSubmit;
+      conf.put(TEST_MOCK_SPARK_SUBMIT.key(), "true");
+
+      client = new LivyClientBuilder(false).setURI(new URI("rsc:/"))
+              .setAll(conf)
+              .build();
+
+      client.stop(true);
+
+      // The FutureListener that calls child.kill() is running in a different thread.
+      // To avoid race condition, add a retry block around verify.
+      DateTime deadline = new DateTime().plusSeconds(10);
```

Comment:
Seems like this would be simpler if you used the `CountDownLatch` approach you used for `waitFor`. Just set up an answer for the `destroy` method too, triggering a different latch instance.